Pattern Creator: Set up creating/editing permissions for logged in wp.org users

coreymckrill opened the pull request:
Enables logged in wp.org users to create and edit patterns on the front end of the Pattern Directory without having to be an actual member of the patterns site.

First I approached this in the way suggested in #30, modifying the permissions checks in the rest controller for the patterns post type. However, it became apparent that the capability checks that were preventing non-site members from creating patterns were used in other places besides the rest controller, so I decided to switch gears and grant the necessary capabilities in certain limited circumstances. Thus the rest controller is not modified at all here.

This also fixes an issue with the pattern theme's stylesheet that was preventing it from loading correctly when trying to access the New Pattern screen while logged out.

Fixes #30

### How to test the changes in this Pull Request:

1. For testing in the local dev environment, create a new user with the Subscriber role and log into the site as that user. This simulates being a wp.org user who is not a member of the site, but who is logged in.
2. Go to the [New Pattern screen](http://localhost:8888/new-pattern/). The editor should load with no errors in the console or broken XHR requests.
3. Create and publish a new pattern. This should succeed and there should be no console errors.
4. Try going to [WP Admin](http://localhost:8888/wp-admin/) and see if you have access to the Block Pattern menu or any of the screens within it. You shouldn't.
5. Try to find scenarios where the Subscriber user has access that it shouldn't.
5. Now log out and log back in as an admin. You should have full access to block patterns in WP Admin and pattern creation on the front end.

<!-- If you can, add the appropriate [Component] label(s). -->

coreymckrill committed:
Ensure theme's stylesheet actually gets enqueued

coreymckrill committed:
More granular permissions checks in editor template

coreymckrill committed:
Grant all logged in users caps for creating patterns in some circumstances

coreymckrill committed:
Fix mapping for custom cap type, add more caps for front end pattern creation

ryelle commented:
This is great! I can't find any issues when using a subscriber role, I also tested a few places with admin and contributor.

I noticed is that if the user tries to edit a pattern they don't own (`/pattern/[ID]/edit/`), it doesn't load anything:

![Screen Shot 2021-07-01 at 16 50 57](https://user-images.githubusercontent.com/541093/124187917-85343580-da8c-11eb-9283-7842d992dc7f.png)

And on the logged out screen, the header styles are a little off (same in the above screenshot too). If this isn't an easy fix, it's fine to leave for another PR.

![Screen Shot 2021-07-01 at 16 54 07](https://user-images.githubusercontent.com/541093/124188218-f673e880-da8c-11eb-8cc3-d83f91699405.png)

coreymckrill commented:
> I noticed is that if the user tries to edit a pattern they don't own (/pattern/[ID]/edit/), it doesn't load anything

Interesting. I'm guessing this is because WP doesn't consider the `/pattern/[ID]/edit/` URI to be a `singular` view of that pattern. We might have to come up with a different conditional to use in the editor template...

coreymckrill committed:
Creator: Editor template tweaks

* Use `is_editing_pattern` for conditional instead of `is_singular`
* Use custom query var to get the correct pattern post ID
* Ensure correct redirect URL for login

coreymckrill committed:
Simplify logic/readability in rewrite rule

coreymckrill commented:
Ok, this should be fixed with 2d36ceb

coreymckrill commented:
> And on the logged out screen, the header styles are a little off (same in the above screenshot too).

I'm not seeing specifically how they are off. But I didn't actually change anything in the stylesheet here, just how it was enqueued, so probably better to save that for a separate PR.

ryelle commented on public_html/wp-content/plugins/pattern-creator/view/editor.php:
```diff
@@ -10,35 +10,50 @@
 
 // @todo Permissions TBD, see https://github.com/WordPress/pattern-directory/issues/30.
```
Can you remove this todo comment before merging?

coreymckrill replied:
Done in 7099169

ryelle reviewed (approved):
All working correctly now 👍🏻

coreymckrill committed:
Remove obsolete code comment

coreymckrill committed:
Fix whitespace

coreymckrill committed:
Remove unnecessary cap assignment

`create_posts` gets mapped to `edit_posts` so there's no need to
set it separately.

coreymckrill committed:
Add unit tests for capabilities
